UserCode/CodeGPU/plugins/KernelManager.cu

```diff
@@ -0,0 +1,134 @@
+#include <cuda.h>
+#include <cuda_runtime.h>
+#include <inttypes.h>
+#include "KernelManager.h"
+#include "HGCalRecHitKernelImpl.cuh"
+
+dim3 nblocks_;
+constexpr dim3 nthreads_(256); //some kernels will potentially not allocate shared memory properly with a lower number
+
+KernelManagerHGCalRecHit::KernelManagerHGCalRecHit(KernelModifiableData<HGCUncalibratedRecHitSoA, HGCRecHitSoA> *data):
+  data_(data)
+{
+  nblocks_ = (data_->nhits + nthreads_.x - 1) / nthreads_.x;
+  nbytes_host_ = (data_->h_out)->nbytes * data_->stride;
+  nbytes_device_ = (data_->d_1)->nbytes * data_->stride;
+}
+
+KernelManagerHGCalRecHit::~KernelManagerHGCalRecHit()
+{
+}
+
+void KernelManagerHGCalRecHit::assign_and_transfer_to_device_()
+{
+  cudaCheck( cudaMemcpyAsync((data_->d_1)->amplitude, (data_->h_in)->amplitude, nbytes_device_, cudaMemcpyHostToDevice) );
+  after_();
+}
+
+void KernelManagerHGCalRecHit::assign_and_transfer_to_device_(const KernelConstantData<HGCeeUncalibratedRecHitConstantData> *h_kcdata, KernelConstantData<HGCeeUncalibratedRecHitConstantData> *d_kcdata)
+{
+  cudaCheck( cudaMemcpyAsync( d_kcdata->data.hgcEE_fCPerMIP_, h_kcdata->data.hgcEE_fCPerMIP_, h_kcdata->data.nbytes, cudaMemcpyHostToDevice) );
+  after_();
+}
+
+void KernelManagerHGCalRecHit::assign_and_transfer_to_device_(const KernelConstantData<HGChefUncalibratedRecHitConstantData> *h_kcdata, KernelConstantData<HGChefUncalibratedRecHitConstantData> *d_kcdata)
+{
+  cudaCheck( cudaMemcpyAsync( d_kcdata->data.hgcHEF_fCPerMIP_, h_kcdata->data.hgcHEF_fCPerMIP_, h_kcdata->data.nbytes, cudaMemcpyHostToDevice) );
+  after_();
+}
+
+void KernelManagerHGCalRecHit::assign_and_transfer_to_device_(const KernelConstantData<HGChebUncalibratedRecHitConstantData> *h_kcdata, KernelConstantData<HGChebUncalibratedRecHitConstantData> *d_kcdata)
+{
+  cudaCheck( cudaMemcpyAsync( d_kcdata->data.weights_, h_kcdata->data.weights_, h_kcdata->data.nbytes, cudaMemcpyHostToDevice) );
+  after_();
+}
+
+void KernelManagerHGCalRecHit::transfer_to_host_and_synchronize_()
+{
+  cudaCheck( cudaMemcpyAsync((data_->h_out)->energy, (data_->d_out)->energy, nbytes_host_, cudaMemcpyDeviceToHost) );
+  after_();
+}
+
+void KernelManagerHGCalRecHit::reuse_device_pointers_()
+{
+  std::swap(data_->d_1, data_->d_2); 
+  after_();
+}
+
+int KernelManagerHGCalRecHit::get_shared_memory_size_(const int& nd, const int& nf, const int& nu, const int& ni, const int& nb) {
+  int dmem = nd*sizeof(double);
+  int fmem = nf*sizeof(float);
+  int umem = nu*sizeof(uint32_t);
+  int imem = ni*sizeof(int);
+  int bmem = nb*sizeof(bool);
+  return dmem + fmem + umem + imem + bmem;
+}
+
+void KernelManagerHGCalRecHit::run_kernels(const KernelConstantData<HGCeeUncalibratedRecHitConstantData> *h_kcdata, KernelConstantData<HGCeeUncalibratedRecHitConstantData> *d_kcdata)
+{
+  assign_and_transfer_to_device_(h_kcdata, d_kcdata);
+  assign_and_transfer_to_device_();
```

Comment:
create two functions with different names instead of overloading the same function